### PR TITLE
[Network] Add simple token bucket rate limiter implementation.

### PR DESCRIPTION
**Pull request by JoshLind**
## Description
This PR adds a simple token bucket rate limiter implementation. This will be used by the internal and external network rate limiters (coming in subsequent PRs).

## Testing Plan
Existing test infrastructure.

---

> [!NOTE]
> **Low Risk**
> Purely additive new utility crate plus tests and workspace wiring; no existing runtime behavior changes unless adopted by follow-up PRs.
> 
> **Overview**
> Adds a new workspace crate, `aptos-token-bucket`, providing a simple token-bucket rate limiter (`TokenBucket` and `SharedTokenBucket`) with time-based refilling via `aptos-time-service`.
> 
> The implementation exposes an all-or-nothing `try_acquire_all()` API that returns either success or the `Instant` when enough tokens will be available (or `None` if the request exceeds capacity), and includes extensive unit tests for refill timing, drift handling, and saturation/edge cases.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 34dd98c67e2d3cdbc4f820ef4dde226c61f9fbc5. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: crates/aptos-token-bucket/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,733 @@
...
+    /// Current number of available tokens
+    tokens: u64,
+    /// Maximum capacity of the bucket
+    capacity: u64,
+    /// Refill rate in tokens per second
+    refill_rate: u64,
+    /// Last time the bucket was refilled
+    last_refill_time: Instant,
+    /// Time service for getting the current time
+    time_service: TimeService,
```
Comment (gregnazario) [inline crates/aptos-token-bucket/src/lib.rs:33]: I also wrote one of these before in the crates/ back at Diem.  It seems fine, though very simple / you can't have split rate limiting.

File: crates/aptos-token-bucket/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,733 @@
...
+        if self.tokens >= requested {
+            // Deduct the tokens
+            self.deduct_tokens(requested);
+
+            Ok(())
+        } else {
+            // Not enough tokens available, return when they will be ready
+            Err(self.time_of_tokens_needed(requested))
+        }
+    }
```
Comment (gregnazario) [inline crates/aptos-token-bucket/src/lib.rs:117]: If you try to acquire all, you should return the number of tokens acquired right?

File: crates/aptos-token-bucket/src/lib.rs (hunk 1 above)
Comment (gregnazario) [inline crates/aptos-token-bucket/src/lib.rs:33]: I wonder if we just use an atomic U64 for tokens, and then it's a best effort refill time.

Wonder how much contention would be there from a mutex rather than the atomic by itself

Comment (JoshLind) [reply crates/aptos-token-bucket/src/lib.rs:33]: Yeah, all we need is something super simple now. We're implementing inbound and outbound network rate limiting. Similar to what we had in the old days, but there's a few different requirements, e.g., per peer and application rate limits for both bytes and messages (so we'll have a token bucket for each unique constraint) 😄

File: crates/aptos-token-bucket/src/lib.rs (hunk 2 above)
Comment (JoshLind) [reply crates/aptos-token-bucket/src/lib.rs:117]: Aah, if you read the function comment, this function is all or nothing, i.e.,:
```
/// Returns successfully if all tokens were acquired, or the time when they
/// will be ready if not enough tokens are available now.
```

File: crates/aptos-token-bucket/src/lib.rs (hunk 1 above)
Comment (JoshLind) [reply crates/aptos-token-bucket/src/lib.rs:33]: We should be okay with a mutex for now 😄 I'd like to get the rate limiters added first, and then we can benchmark them to see how they perform. But yeah, if it's a problem, maybe we can optimize with an atomic or something smarter 🚀
